File: common/src/main/java/juuxel/adorn/item/CarpetedTopPlacementContext.java
package juuxel.adorn.item;

import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.item.context.UseOnContext;

public final class CarpetedTopPlacementContext extends BlockPlaceContext {
    public CarpetedTopPlacementContext(UseOnContext context) {
        super(context);
        // We know that the block is a carpet block
        replaceClicked = true;
    }
}

File: src/main/kotlin/juuxel/adorn/item/CarpetedTopPlacementContext.kt
package juuxel.adorn.item

import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.ItemUsageContext
import net.minecraft.util.hit.BlockHitResult

/**
 * An `ItemPlacementContext` subclass for placing blocks inside carpets.
 */
class CarpetedTopPlacementContext(parent: ItemUsageContext) : ItemPlacementContext(
    parent.world,
    parent.player,
    parent.hand,
    parent.stack,
    BlockHitResult(parent.hitPos, parent.side, parent.blockPos, parent.method_17699())
) {
    init {
        // We know that the block is a carpet block
        canReplaceExisting = true
    }
}
